File: src/chat/chat.service.ts
import { Injectable, InternalServerErrorException, Logger } from '@nestjs/common';
import { Chat, Message, Prisma, PrismaClient, User } from '@prisma/client';
import { withAccelerate } from '@prisma/extension-accelerate';
import { ErrorCodes } from 'enums/error-codes.enum';
import { CreatChatDTO } from './dto/create-chat.dto';

@Injectable()
export class ChatService {
  private readonly prisma = new PrismaClient().$extends(withAccelerate());
  private readonly logger = new Logger(ChatService.name);

  createChat(userId: User['id'], createChatDto: CreatChatDTO): Promise<Chat> {
    this.logger.log(`Creating chat with name: ${createChatDto.name} for user with id: ${userId}`);
    return this.prisma.$transaction(tx => {
      try {
        return tx.chat.create({
          data: {
            name: createChatDto.name,
            picture: createChatDto.picture,
            description: createChatDto.description,
            users: {
              createMany: { data: createChatDto.users.map(id => ({ userId: id })) },
            },
            userKeys: {
              createMany: {
                data: createChatDto.userKeys.map(userKey => ({
                  key: userKey.key,
                  userId: userKey.userId,
                })),
              },
            },
          },
        });
      } catch (err) {
        this.logger.error(`Failed to create chat. Error: ${err}`);
        throw new InternalServerErrorException({ error: ErrorCodes.API_ERROR });
      }
    });
  }

  getChatPreviews(userId: User['id']): Promise<
    Prisma.ChatGetPayload<{
      include: {
        messages: {
          select: {
            content: true;
          };
        };
      };
    }>[]
  > {
    this.logger.log(`Finding chats for user with id: ${userId}`);
    return this.prisma.chat.findMany({
      where: {
        users: {
          some: {
            userId,
          },
        },
      },
      include: {
        messages: {
          select: {
            content: true,
          },
          orderBy: {
            createdAt: 'desc',
          },
          take: 1,
        },
      },
    });
  }

  getChatById(
    userId: User['id'],
    chatId: Chat['id'],
  ): Promise<
    Prisma.ChatGetPayload<{
      include: {
        messages: {
          select: {
            id: true;
            createdAt: true;
            updatedAt: true;
            content: true;
            iv: true;
            sender: {
              select: {
                id: true;
                name: true;
              };
            };
          };
        };
        users: {
          select: {
            user: {
              select: {
                id: true;
                name: true;
                profilePicture: true;
              };
            };
          };
        };
        userKeys: {
          select: {
            key: true;
            userId: true;
          };
        };
      };
    }>
  > {
    this.logger.log(`Finding chat with id: ${chatId} for user with id: ${userId}`);
    let chat;
    try {
      chat = this.prisma.chat.findUnique({
        where: {
          id: chatId,
        },
        include: {
          messages: {
            select: {
              id: true,
              createdAt: true,
              updatedAt: true,
              content: true,
              iv: true,
              sender: {
                select: {
                  id: true,
                  name: true,
                },
              },
            },
            orderBy: {
              createdAt: 'desc',
            },
          },
          users: {
            select: {
              user: {
                select: {
                  id: true,
                  name: true,
                  profilePicture: true,
                },
              },
            },
          },
          userKeys: {
            select: {
              key: true,
              userId: true,
            },
          },
        },
      });
    } catch (err) {
      this.logger.error(`Failed to find chat with id: ${chatId} for user with id: ${userId}. Error: ${err}`);
      throw new InternalServerErrorException({ error: ErrorCodes.API_ERROR });
    }
    if (!chat) {
      this.logger.error(`Failed to find chat with id: ${chatId} for user with id: ${userId}`);
      throw new InternalServerErrorException({ error: ErrorCodes.NOT_FOUND });
    }
    return chat;
  }

  sendMessage(
    senderId: User['id'],
    chatId: Chat['id'],
    content: Message['content'],
    iv: Message['iv'],
  ): Promise<
    Prisma.MessageGetPayload<{
      include: {
        sender: {
          select: {
            id: true;
            name: true;
          };
        };
      };
    }>
  > {
    this.logger.log(`Sending message for user with id: ${senderId} in chat with id: ${chatId}`);
    return this.prisma.$transaction(async tx => {
      try {
        const chat = await tx.chat.findUnique({
          where: {
            id: chatId,
          },
        });

        if (!chat) {
          this.logger.error(`Failed to send message for user with id: ${senderId} in chat with id: ${chatId}`);
          throw new InternalServerErrorException({ error: ErrorCodes.NOT_FOUND });
        }

        return tx.message.create({
          data: {
            content,
            senderId,
            chatId,
            iv,
          },
          include: {
            sender: {
              select: {
                id: true,
                name: true,
              },
            },
          },
        });
      } catch (err) {
        this.logger.error(
          `Failed to send message for user with id: ${senderId} in chat with id: ${chatId}. Error: ${err}`,
        );
        throw new InternalServerErrorException({ error: ErrorCodes.API_ERROR });
      }
    });
  }
}

File: src/api/chat/chat.service.go
package chat

import (
	"easyflow-backend/src/api"
	"easyflow-backend/src/api/auth"
	"easyflow-backend/src/common"
	"easyflow-backend/src/database"
	"easyflow-backend/src/enum"
	"net/http"

	"gorm.io/gorm"
)

func CreateChat(db *gorm.DB, payload *CreateChatRequest, jwtPayload *auth.JWTPayload, logger *common.Logger) (*CreateChatResponse, *api.ApiError) {
	var users []database.User
	var userKeys []UserKeyEntry

	// Start a transaction
	tx := db.Begin()
	logger.Printf("Attempting to create chat with name: %s", payload.Name)

	//get usres from payload.UserKeys
	for _, userKey := range payload.UserKeys {
		user := database.User{}
		if err := tx.Where("id = ?", userKey.UserID).First(&user).Error; err != nil {
			tx.Rollback()
			logger.PrintfError("Error getting user with id: %s", userKey.UserID)
			return nil, &api.ApiError{
				Code:  http.StatusInternalServerError,
				Error: enum.ApiError,
			}
		}
		users = append(users, user)
		userKeys = append(userKeys, userKey)
	}

	if len(users) != len(payload.UserKeys) {
		tx.Rollback()
		logger.PrintfError("User keys and users length mismatch")
		return nil, &api.ApiError{
			Code:  http.StatusNotFound,
			Error: enum.UserNotFound,
		}
	}

	if len(users) != len(userKeys) {
		tx.Rollback()
		logger.PrintfError("User keys and users length mismatch")
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	chat := &database.Chat{
		Name:        payload.Name,
		Picture:     payload.Picture,
		Description: payload.Description,
		Messages:    nil,
	}

	if err := tx.Create(chat).Error; err != nil {
		tx.Rollback()
		logger.PrintfError("Error creating chat: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	for i, user := range users {
		chatUserKeys := &database.ChatUserKeys{
			ChatId: chat.Id,
			UserId: user.Id,
			Key:    userKeys[i].Key,
		}

		if err := tx.Create(chatUserKeys).Error; err != nil {
			tx.Rollback()
			logger.PrintfError("Error creating chat user key: %s", err)
			return nil, &api.ApiError{
				Code:  http.StatusInternalServerError,
				Error: enum.ApiError,
			}
		}
	}

	if err := tx.Commit().Error; err != nil {
		logger.PrintfError("Error committing transaction: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	return &CreateChatResponse{
		Id:          chat.Id,
		CreatedAt:   chat.CreatedAt.String(),
		UpdateAt:    chat.UpdatedAt.String(),
		Name:        chat.Name,
		Picture:     chat.Picture,
		Description: chat.Description,
	}, nil
}

func GetChatPreviews(db *gorm.DB, jwtPayload *auth.JWTPayload, logger *common.Logger) ([]GetChatPreviewResponse, *api.ApiError) {
	logger.Printf("Attempting to get chat previews for user: %s", jwtPayload.UserId)
	var chatUserKeys []database.ChatUserKeys
	chatPreviews := []GetChatPreviewResponse{}

	if err := db.Where("user_id = ?", jwtPayload.UserId).Find(&chatUserKeys).Error; err != nil {
		logger.Printf("Error getting chats for user: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	for _, chatUserKey := range chatUserKeys {
		var chat database.Chat
		if err := db.Where("id = ?", chatUserKey.ChatId).First(&chat).Error; err != nil {
			logger.Printf("Error getting chat with id: %s. %s", chatUserKey.ChatId, err)
			return nil, &api.ApiError{
				Code:  http.StatusInternalServerError,
				Error: enum.ApiError,
			}
		}

		var lastMessage *database.Message = nil
		if err := db.Where("chat_id = ?", chatUserKey.ChatId).Order("created_at desc").First(&lastMessage).Error; err != nil {
			if err != gorm.ErrRecordNotFound {
				// If there's another error, log it and return
				logger.PrintfError("Error getting last message for chat with id: %s. Error: %s", chatUserKey.ChatId, err.Error())
				return nil, &api.ApiError{
					Code:  http.StatusInternalServerError,
					Error: enum.ApiError,
				}
			}

			chatPreview := GetChatPreviewResponse{
				CreateChatResponse: CreateChatResponse{
					Id:          chat.Id,
					CreatedAt:   chat.CreatedAt.String(),
					UpdateAt:    chat.UpdatedAt.String(),
					Name:        chat.Name,
					Picture:     chat.Picture,
					Description: chat.Description,
				},
				LastMessage: &lastMessage.Content,
			}

			chatPreviews = append(chatPreviews, chatPreview)
		}
	}

	return chatPreviews, nil
}

func GetChatById(db *gorm.DB, chatId string, jwtPayload *auth.JWTPayload, logger *common.Logger) (*GetChatByIdResponse, *api.ApiError) {
	var chat database.Chat
	if err := db.Where("id = ?", chatId).First(&chat).Error; err != nil {
		logger.Printf("Error getting chat with id: %s. Error: %s", chatId, err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	var chatUserKeys []database.ChatUserKeys
	if err := db.Where("chat_id = ? AND user_id = ?", chatId, jwtPayload.UserId).Find(&chatUserKeys).Error; err != nil {
		logger.Printf("Error getting chat user key for chat with id: %s. Error: %s", chatId, err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	var Messages []database.Message
	if err := db.Where("chat_id = ?", chatId).Order("created_at desc").Find(&Messages).Limit(50).Error; err != nil {
		logger.Printf("Error getting messages for chat with id: %s. Error: %s", chatId, err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	// Mappings
	usersEntries := []UserEntry{}
	for _, chatUserKey := range chatUserKeys {
		var user database.User
		if err := db.Where("id = ?", chatUserKey.UserId).First(&user).Error; err != nil {
			logger.Printf("Error getting user with id: %s. Error: %s", chatUserKey.UserId, err)
			return nil, &api.ApiError{
				Code:  http.StatusInternalServerError,
				Error: enum.ApiError,
			}
		}

		usersEntries = append(usersEntries,
			UserEntry{
				Id:   user.Id,
				Name: user.Name,
				Bio:  user.Bio,
			},
		)
	}

	// TODO: Just make one object for user keys not array
	userKeyEntries := []UserKeyEntry{}
	for _, chatUserKey := range chatUserKeys {
		userKeyEntries = append(userKeyEntries,
			UserKeyEntry{
				UserID: chatUserKey.UserId,
				Key:    chatUserKey.Key,
			},
		)
	}

	messageEntries := []MessageEntry{}
	for _, message := range Messages {
		messageEntries = append(messageEntries,
			MessageEntry{
				Id:        message.Id,
				CreatedAt: message.CreatedAt.String(),
				UpdatedAt: message.UpdatedAt.String(),
				Content:   message.Content,
				Iv:        message.Iv,
				SenderId:  message.SenderId,
			},
		)
	}

	return &GetChatByIdResponse{
		CreateChatResponse: CreateChatResponse{
			Id:          chat.Id,
			CreatedAt:   chat.CreatedAt.String(),
			UpdateAt:    chat.UpdatedAt.String(),
			Name:        chat.Name,
			Picture:     chat.Picture,
			Description: chat.Description,
		},
		Users:    usersEntries,
		UserKeys: userKeyEntries,
		Messages: messageEntries,
	}, nil

}
